File: python/238_Product_Of_Array_Except_Self.py
class Product_Of_Array_Except_Self:
  def productExceptSelf(self, nums: list[int]) -> list[int]:
    n = len(nums)
    result = [1] * n
    for i in range(1,n):
      result[i] = nums[i - 1] * result[i - 1]
    product = 1
    for i in range(n - 1, -1, -1):
      result[i] = result[i] * product
      product *= nums[i]
    return result

File: c++/238_Product_Of_Array_Except_Self.cpp
#include <vector>

using namespace std;

class Product_Of_Array_Except_Self {
public:
  vector<int> productExceptSelf(vector<int>& nums) {
    vector<int> answer(nums.size());
    answer[0] = 1;
    for (int i = 1; i < nums.size(); i++){
      answer[i] = nums[i-1] * answer[i-1];
    }
    int product = 1;
    for (int i = (nums.size() - 2); i >= 0; i--){
      product *= nums[i+1];
      answer[i] *= product;
    }
    return answer;
  }
};